nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java

```diff
@@ -0,0 +1,502 @@
+
+                    } else {
+                        final Object value = parseUnknownField(subStartElement);
+                        if (value != null) {
+                            putUnknownTypeInMap(recordValues, fieldName, value);
+                        }
+                    }
+                }
+            } else if (xmlEvent.isEndElement()) {
+                break;
+            } else if (xmlEvent.isCharacters()) {
+                final Characters characters = xmlEvent.asCharacters();
+                if (!characters.isWhiteSpace()) {
+                    if (contentFieldName != null) {
+                        final Optional<RecordField> field = schema.getField(contentFieldName);
+                        if (field.isPresent()) {
+                            Object value = parseCharacterForType(characters, contentFieldName, field.get().getDataType());
+                            recordValues.put(contentFieldName, value);
+                        }
+                    } else {
+                        logger.debug("Found content for field that is defined as record but property \"Field Name for Content\" is not set. " +
+                                "The content will not be added to record.");
+                    }
+                }
+            }
+        }
+        for (final Map.Entry<String,Object> entry : recordValues.entrySet()) {
+            if (entry.getValue() instanceof List) {
+                recordValues.put(entry.getKey(), ((List) entry.getValue()).toArray());
+            }
+        }
+
+        if (recordValues.size() > 0) {
+            return new MapRecord(schema, recordValues);
+        } else {
+            return null;
+        }
+    }
+
+    private void putUnknownTypeInMap(Map<String, Object> values, String fieldName, Object fieldValue) {
+        final Object oldValues = values.get(fieldName);
+
+        if (oldValues != null) {
+            if (oldValues instanceof List) {
+                ((List) oldValues).add(fieldValue);
+            } else {
+                values.put(fieldName, new ArrayList<Object>(){{ add(oldValues); add(fieldValue); }});
+            }
+        } else {
+            values.put(fieldName, fieldValue);
+        }
+    }
+
+    private Object parseAttributeForType(Attribute attribute, String fieldName, DataType dataType) {
+        switch (dataType.getFieldType()) {
+            case BOOLEAN:
+            case BYTE:
+            case CHAR:
+            case DOUBLE:
+            case FLOAT:
+            case INT:
+            case LONG:
+            case SHORT:
+            case STRING:
+            case DATE:
+            case TIME:
+            case TIMESTAMP: {
+                return DataTypeUtils.convertType(attribute.getValue(), dataType, LAZY_DATE_FORMAT, LAZY_TIME_FORMAT, LAZY_TIMESTAMP_FORMAT, fieldName);
+            }
+        }
+        return null;
+    }
+
+    private Object parseCharacterForType(Characters character, String fieldName, DataType dataType) {

```

Comment:
Probably can create a single parseStringDataForType(String data, String fieldName, DataType dataType) method and call that from both parseAttributeForType and parseCharacterForType, rather than having to repeat the large case statement.